components/patient-data/impl/src/main/java/org/phenotips/data/internal/PatientOnCancelDeleter.java

```diff
@@ -0,0 +1,129 @@
+ */
+package org.phenotips.data.internal;
+
+import org.phenotips.data.Patient;
+import org.phenotips.data.events.PatientCancelledEvent;
+import org.phenotips.data.events.PatientEvent;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.model.reference.EntityReferenceResolver;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+
+import java.io.IOException;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.criteria.impl.RevisionCriteria;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.doc.rcs.XWikiRCSNodeInfo;
+
+/**
+ * Update the Date of birth aggregated property whenever the birth year, month and day properties are specified in the
+ * request.
+ *
+ * @version $Id$
+ */
+@Component
+@Named("patient-oncancel-deleter")
+@Singleton
+public class PatientOnCancelDeleter extends AbstractEventListener
+{
+    private static final String AUTOSAVE_COMMENT = "(Autosaved)";
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Logger logger;
+
+    /** Fills in missing reference fields with those from the current context document to create a full reference. */
+    @Inject
+    @Named("current")
+    private EntityReferenceResolver<EntityReference> currentResolver;
+
+
+    /** Default constructor, sets up the listener name and the list of events to subscribe to. */
+    public PatientOnCancelDeleter()
+    {
+        super("patient-oncancel-deleter", new PatientCancelledEvent());
+    }
+
+    @Override
+    public void onEvent(final Event event, final Object source, final Object data)
+    {
+        Patient patient = ((PatientEvent) event).getPatient();
+        XWikiContext context = contextProvider.get();
+        XWiki wiki = context.getWiki();
+        try {
+            XWikiDocument doc = wiki.getDocument(patient.getDocument(), context);
+            if (isNew(doc, context)) {
+                wiki.deleteDocument(doc, true, context);
+
+                EntityReference defaultReference = new EntityReference(wiki.getDefaultPage(context),
+                    EntityType.DOCUMENT, new EntityReference(wiki.getDefaultSpace(context), EntityType.SPACE));
+                defaultReference = currentResolver.resolve(defaultReference, EntityType.DOCUMENT);
+                String url = wiki.getURL(new DocumentReference(defaultReference), "view", context);
+                if (context.getResponse().isCommitted()) {
+                    context.getResponse().setHeader("Location", url);
```

Comment:
Actually the problem is that the `ActionExecutedEvent` is sent **after** the response has been processed, you should listen to `ActionExecutingEvent` instead.